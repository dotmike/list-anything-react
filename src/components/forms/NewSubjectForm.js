import React from "react";
import PropTypes from "prop-types";
import { Form, Button, Message, Segment } from "semantic-ui-react";

import InlineError from "../messages/InlineError";
import TabsAccordionForm from "./TabsAccordionForm";
import FieldsAccordionForm from "./FieldsAccordionForm";

class NewSubjectForm extends React.Component {
  state = {
    data: {
      description: "",
      tabs: [],
      fields: []
    },
    loading: false,
    errors: {}
  };

  onChange = e =>
    this.setState({
      data: { ...this.state.data, [e.target.name]: e.target.value }
    });

  onSubmit = () => {
    const errors = this.validate(this.state.data);
    this.setState({ errors });
    if (Object.keys(errors).length === 0) {
      this.setState({ loading: true });
      this.props.submit(this.state.data).catch(err => {
        if (err.response.status === 500)
          this.setState({
            errors: { global: "Internal server error" },
            loading: false
          });
        else
          this.setState({ errors: err.response.data.errors, loading: false });
      });
    }
  };

  addTab = tab => {
    const newTab = { ...tab };
    this.setState(prevState => ({
      data: {
        ...prevState.data,
        tabs: [...prevState.data.tabs, newTab]
      }
    }));
  };

  removeTab = item => {
    this.setState(prevState => ({
      data: {
        ...prevState.data,
        tabs: prevState.data.tabs.filter(elem => elem !== item)
      }
    }));
  };

  addField = field => {
    const newField = { ...field };
    this.setState(prevState => ({
      data: {
        ...prevState.data,
        fields: [...prevState.data.fields, newField]
      }
    }));
  };

  removeField = item => {
    this.setState(prevState => ({
      data: {
        ...prevState.data,
        fields: prevState.data.fields.filter(elem => elem !== item)
      }
    }));
  };

  validate = data => {
    const errors = {};
    if (!data.description) errors.description = "Can't be blank";
    if (data.tabs.length === 0) errors.tabs = "Needs at least one tab";
    if (data.fields.length === 0) errors.fields = "Needs at least one field";
    return errors;
  };

  render() {
    const { data, errors, loading } = this.state;

    return (
      <Form onSubmit={this.onSubmit} loading={loading} size="large">
        {errors.global && (
          <Message negative>
            <Message.Header>Something went wrong</Message.Header>
            <p>{errors.global}</p>
          </Message>
        )}
        <Segment stacked textAlign="left">
          <Form.Field error={!!errors.description}>
            <Form.Input
              fluid
              type="text"
              placeholder="subject description"
              value={data.description}
              onChange={this.onChange}
              name="description"
            />
            {errors.description && <InlineError text={errors.description} />}
          </Form.Field>

          <TabsAccordionForm
            tabs={data.tabs}
            addTab={this.addTab}
            removeTab={this.removeTab}
          />
          {errors.tabs && <InlineError text={errors.tabs} />}

          <FieldsAccordionForm
            fields={data.fields}
            addField={this.addField}
            removeField={this.removeField}
          />
          {errors.fields && <InlineError text={errors.fields} />}

          <Button color="teal" fluid size="large">
            Create
          </Button>
        </Segment>
      </Form>
    );
  }
}

NewSubjectForm.propTypes = {
  submit: PropTypes.func.isRequired
};

export default NewSubjectForm;
