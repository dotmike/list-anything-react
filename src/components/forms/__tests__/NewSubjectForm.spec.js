import React from "react";
import { shallow } from "enzyme";

import NewSubjectForm from "../NewSubjectForm";

const defaultProps = {
  submit: () => {}
};

describe("NewSubjectForm", () => {
  it("renders correctly", () => {
    const wrapper = shallow(<NewSubjectForm {...defaultProps} />);
    expect(wrapper).toMatchSnapshot();
  });
});
