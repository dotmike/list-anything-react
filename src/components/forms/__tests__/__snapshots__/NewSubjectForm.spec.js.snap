// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewSubjectForm renders correctly 1`] = `
<Form
  as="form"
  error={false}
  loading={false}
  onSubmit={[Function]}
  size="large"
>
  <Segment
    stacked={true}
    textAlign="left"
  >
    <FormField
      error={false}
    >
      <FormInput
        as={[Function]}
        control={[Function]}
        fluid={true}
        id="subject-description-input"
        name="description"
        onChange={[Function]}
        placeholder="subject description"
        type="text"
        value=""
      />
    </FormField>
    <TabsAccordionForm
      addTab={[Function]}
      removeTab={[Function]}
      tabs={Array []}
    />
    <FieldsAccordionForm
      addField={[Function]}
      fields={Array []}
      removeField={[Function]}
    />
    <Divider />
    <Button
      as="button"
      color="teal"
      fluid={true}
      size="large"
    >
      Create
    </Button>
  </Segment>
</Form>
`;
